### ISSUE TITLE
Unity 2019.3

### ISSUE THREAD

**Issue by craigktreasure** [Azure/azure-spatial-anchors-samples#149]
Hey everyone,

As you may know, Azure Spatial Anchors supports many platforms and technology stacks. Unfortunately, Unity 2019.3 comes with some breaking changes that prevent us from being able to support all of those platforms (specifically HoloLens in this case) on that version of Unity. We're actively having conversations with our friends at Unity to find a solution, but in the meantime, we need to reiterate that we currently only support [Unity 2019.1 and Unity 2019.2](https://docs.microsoft.com/en-us/azure/spatial-anchors/quickstarts/get-started-unity-ios#unity-20193).

Having said that, we also recognize that there are many of you who are ambitious, need new features and fixes, and may not need to support all of the platforms that Azure Spatial Anchors does. Below is a guide to update the SDK for Android and iOS. Just remember that there is no support from us for Unity 2019.3 at this time.

Thanks!

# Updating Azure Spatial Anchors Unity SDK for Unity 2019.3

1. Open the project in Unity 2019.3.
2. You'll be asked if you want to upgrade the project to a new version of Unity. Select **CONFIRM**.
3. You'll be asked if you want to update the asset database to version 2. Select **YES**.

## Updating to AR Foundation 3

1. Open the Package Manager (**Window -> Package Manager**) and update the following packages:
   1. AR Foundation: 3.0.1
   2. AR Subsystems: 3.0.0
   3. ARCore XR Plugin: 3.0.1
   4. ARKit XR Plugin: 3.0.1
2. Update the ASA SDK to ARFoundation 3 objects:
   1. In `Assets\AzureSpatialAnchors.SDK\Scripts\SpatialAnchorManager.cs`:
      1. Replace `ARReferencePoint` references with `ARAnchor`.
      2. Replace `ARReferencePointManager` references with `ARAnchorManager`.
      3. Replace `ARReferencePointsChangedEventArgs` references with `ARAnchorsChangedEventArgs`.
      4. Replace the `referencePointsChanged` event reference on `arReferencePointManager` with `anchorsChanged`.
   2. In `Assets\AzureSpatialAnchors.SDK\Plugins\ARFoundation\AnchorHelpers.cs`:
      1. Replace `ARReferencePoint` references with `ARAnchor`.
      2. Replace the `AddReferencePoint` method call on `arReferencePointManager` with `AddAnchor`.
   3. In `Assets\AzureSpatialAnchors.SDK\Plugins\ARFoundation\UnityARFoundationAnchorComponent.cs`:
      1. Replace `ARReferencePoint` references with `ARAnchor`.
      2. Replace the `RemoveReferencePoint` method call on `SpatialAnchorManager.arReferencePointManager` with `RemoveAnchor`.

## Updates required for Unity 2019.3

Follow the instructions for updating the ASA Unity SDK to AR Foundation 3 above.

Follow the instructions below to update the SDK **BEFORE** building\deploying your project from Unity.

### iOS

Update `Assets/AzureSpatialAnchors.SDK/Plugins/iOS/Podfile` with the following content replacing `<<ASA VERSION>>` with the version of Azure Spatial Anchors you wish to use. At the time of writing, `2.2.1` is the latest.

``` ruby
platform :ios, '11.0'
use_frameworks!

target 'Unity-iPhone' do
    target 'UnityFramework' do
        # Pods for UnityFramework
        pod 'AzureSpatialAnchors', '<<ASA VERSION>>'
    end
end
```

### Android

Update `Assets\Plugins\Android\mainTemplate.gradle` with the following content replacing `<<ASA VERSION>>` with the version of Azure Spatial Anchors you wish to use. At the time of writing, `2.2.1` is the latest.

``` groovy
apply plugin: 'com.android.library'
**APPLY_PLUGINS**

def azureSpatialAnchorsSdkVersion = '<<ASA VERSION>>'

/*
The ARCore aar library contains the native shared libraries.  These are
extracted before building to a temporary directory.
 */
def natives_libpath = "${buildDir}/natives"

// Create a configuration to mark which aars to extract .so files from
configurations { natives }

dependencies {
    implementation fileTree(dir: 'libs', include: ['*.jar'])
**DEPS**
    // Azure Spatial Anchors
    implementation ("com.microsoft.azure.spatialanchors:spatialanchors_ndk:[${azureSpatialAnchorsSdkVersion}]") {
        exclude group: 'com.google.ar'
    }
    natives "com.microsoft.azure.spatialanchors:spatialanchors_ndk:[${azureSpatialAnchorsSdkVersion}]"
}

android {
    compileSdkVersion **APIVERSION**
    buildToolsVersion '**BUILDTOOLS**'

    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }

    defaultConfig {
        minSdkVersion **MINSDKVERSION**
        targetSdkVersion **TARGETSDKVERSION**
        ndk {
            abiFilters **ABIFILTERS**
        }
        versionCode **VERSIONCODE**
        versionName '**VERSIONNAME**'
        consumerProguardFiles 'proguard-unity.txt'**USER_PROGUARD**

        externalNativeBuild {
            cmake {
                cppFlags "-std=c++14", "-Wall"
                arguments "-DANDROID_STL=c++_static",
                        "-DNATIVE_LIBPATH=${natives_libpath}/jni",
                        "-DNATIVE_INCLUDE=${project.rootDir}/libraries/include"
            }
        }
    }

    lintOptions {
        abortOnError false
    }

    aaptOptions {
        ignoreAssetsPattern = "!.svn:!.git:!.ds_store:!*.scc:.*:!CVS:!thumbs.db:!picasa.ini:!*~"
    }**PACKAGING_OPTIONS**
}**REPOSITORIES****SOURCE_BUILD_SETUP**
**EXTERNAL_SOURCES**

// Extracts the shared libraries from aars in the natives configuration.
// This is done so that NDK builds can access these libraries.
task extractNativeLibraries() {
    // Always extract, this insures the native libs are updated if the version changes.
    outputs.upToDateWhen { false }
    doFirst {
        configurations.natives.files.each { f ->
            copy {
                from zipTree(f)
                into natives_libpath
                include "jni/**/*"
            }
        }
    }
}

tasks.whenTaskAdded {
    task-> if (task.name.contains("external") && !task.name.contains("Clean")) {
        task.dependsOn(extractNativeLibraries)
    }
}
```

**Comment by rogueyoshi** [comment]
The higher versioned ARFoundation 3 packages (3.1.x, I believe) have update scripts on them to update the entire project without manually having to replace the reference. They automatically run upon installing the package.

**Comment by tomkrikorian** [comment]
Will you support ARFoundation 4 with 2019.3 ?

**Comment by t5ujiri** [comment]
Just noting, you need to update pod version of ASA or you'll get build time error

```rb
platform :ios, '11.0'
use_frameworks!

target 'Unity-iPhone' do
    target 'UnityFramework' do
        # Pods for UnityFramework
        # pod 'AzureSpatialAnchors', '1.3.0'
        pod 'AzureSpatialAnchors', '2.2.1'
    end
end
```

**Comment by midopooler** [comment]
Didn't worked for me!

**Comment by t5ujiri** [comment]
```rb
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '11.0'

target 'Unity-iPhone' do
end

target 'UnityFramework' do
  pod 'AzureSpatialAnchors', '2.2.1'
  # add other dependencies below as static framework
  # e.g. pod 'Firebase/Core', '6.17.0'
end
```
I found that you don't need to make `UnityFramework` as sub-target of `Unity-iPhone` nor specify  `use_framework!`, but have to make `Unity-iPhone` target anyway.

**Comment by spaskhalov** [comment]
Still have "Undefined symbols for architecture arm64:
"_ssc_cloud_spatial_anchor_session_set_log_level", " errors in XCode.

**Comment by darax** [comment]
@spaskhalov, what version of unity are you running?  What file did you open in xcode after syncing the pod file?

**Comment by spaskhalov** [comment]
@darax Unity 2019.3.10f1. I open Unity-iPhone.xcworkspace in xcode. By the way version of xcode 11.3.1, if it matters. Content of my Podfile:

```
platform :ios, '11.0'
use_frameworks!
target 'Unity-iPhone' do
    target 'UnityFramework' do
        # Pods for UnityFramework
        pod 'AzureSpatialAnchors', '2.2.1'
    end
end

```

**Comment by spaskhalov** [comment]
I found why our project didn't work. We use Firebase & ["External Dependency Manager for Unity(EDM4U)"](https://github.com/googlesamples/unity-jar-resolver). The podfile from azure `Assets/AzureSpatialAnchors.SDK/Plugins/iOS/Podfile` has been overwritten by EDM4U. 
Thus, in the Podfile inside xcode project, the pod 'AzureSpatialAnchors' was simply not added. I manually made changes to the Podfile inside the xcode project and now it looks like this:
```
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '11.0'

target 'Unity-iPhone' do
end

target 'UnityFramework' do
  pod 'Crashlytics', '3.13.4'
  pod 'Fabric', '1.10.2'
  pod 'Firebase/Analytics', '6.17.0'
  pod 'Firebase/Auth', '6.17.0'
  pod 'Firebase/Core', '6.17.0'
  pod 'Firebase/Database', '6.17.0'
  pod 'Firebase/Firestore', '6.17.0'
  pod 'Firebase/Functions', '6.17.0'
  pod 'Firebase/Messaging', '6.17.0'
  pod 'Firebase/Storage', '6.17.0'
  pod 'AzureSpatialAnchors', '2.2.1'
end
```
I'm not sure whose mistake it is. In older versions of Unity, there was no conflict between Azure & Firebase.

**Comment by faizan1990** [comment]
Does the first line in Podfile mean anything?

`platform :ios, '11.0'`

I am pretty sure no one is running iOS 11 on their iPhones anymore. Please confirm.

Also, I made the changes advised in this thread, but they didn't work for me.

**Comment by demeyera** [comment]
According to the sessions on both Build and Mixed Reality Dev Days, Azure Spatial Anchors is in GA and fully compatible with Unity 2019.3, so please advise on getting it working properly with 2019.3, as this repo seems to be untouched since March 2020..

**Comment by craigktreasure** [comment]
It is true that Azure Spatial Anchors is now GA, however, we still do not support Unity 2019.3. Curious where you heard that. Sorry for any confusion.

**Comment by jgwinner** [comment]
From Neena Kamath - the product lead for Azure Mixed reality, Microsoft, as well as Simon Skaria - microsoft

This was during Microsoft Build. The speaker was definite about it. It might have been another person, I didn't write down exactly who said it, but the talk was given by Neena.

In a few days you can hear them: https://mybuild.microsoft.com/speaker/6a6c93c4-9a0a-4fdc-9f18-d4b5611f9747?source=sessiondetail&sessionId=30c8b2c7-9261-4a50-81a0-b7e6f50b6d37 

        == John ==

**Comment by craigktreasure** [comment]
I'm afraid there's been some confusion and I am sorry about that. I'll sync up with Neena, but at this time Azure Spatial Anchors still does not support Unity 2019.3. We're working on a solution that will allow us to support all of our platforms and partners on Unity 2019.3 and will update everyone here when the time comes.

**Comment by tomkrikorian** [comment]
Hi @craigktreasure,

Just my two cents on the 2019.3 update, it would be amazing if you used the ARAnchor from ARFoundation 4 and not the WorldAnchor as a NativeAnchor. Would make development much easier.

**Comment by craigktreasure** [comment]
Thanks for the suggestion. We'll look into it.

**Comment by xwipeoutx** [comment]
Hi @craigktreasure, any updates on this one? Ideally with a "2019.4" instead of "2019.3" now 😊

Building an app using ASA, targeting both HL2 and mobile, so we're stuck on 2019.2 for now, which isn't ideal.

**Comment by craigktreasure** [comment]
Sorry everyone. We're still working on getting an SDK out that makes everyone happy. We have several moving targets at play.

I have a branch ([craig/unity2019.3](https://github.com/Azure/azure-spatial-anchors-samples/tree/craig/unity2019.3)) that is all setup for 2019.3 using our existing SDK. It should work for all the platforms we currently support. I have not tried it on 2019.4 as that was only recently released. If anyone has a few minutes to try it out, please let me know.

Thanks!

**Comment by tomkrikorian** [comment]
Tried Unity 2019.4 and it works on Android :) 
@craigktreasure : When you say "It should work for all the platforms we currently support.", do you include Hololens or is it still not working ?

**Comment by nooxouille** [comment]
> Tried Unity 2019.4 and it works on Android :)
Super nice!

> @craigktreasure : When you say "It should work for all the platforms we currently support.", do you include Hololens or is it still not working ?

My guess is that it is working for the Hololens, as stated on the branch's page.
I might give it a try, but my agenda is quite loaded so no promises here.

**Comment by spaskhalov** [comment]
> Sorry everyone. We're still working on getting an SDK out that makes everyone happy. We have several moving targets at play.
> 
> I have a branch ([craig/unity2019.3](https://github.com/Azure/azure-spatial-anchors-samples/tree/craig/unity2019.3)) that is all setup for 2019.3 using our existing SDK. It should work for all the platforms we currently support. I have not tried it on 2019.4 as that was only recently released. If anyone has a few minutes to try it out, please let me know.
> 
> Thanks!

I tried to build the app for iOS in Unity 2019.4 and got a lot of errors in XCode. A small piece from the log:

`Undefined symbols for architecture arm64:
  "_ssc_cloud_spatial_anchor_session_stop", referenced from:
      _NativeLibrary_ssc_cloud_spatial_anchor_session_stop_m68322D6DC77508A859AF2EDB61C54A7C6B44B380 in Assembly-CSharp9.o
     (maybe you meant: _NativeLibrary_ssc_cloud_spatial_anchor_session_stop_m68322D6DC77508A859AF2EDB61C54A7C6B44B380)
  "_ssc_session_status_get_ready_for_create_progress", referenced from:
      _NativeLibrary_ssc_session_status_get_ready_for_create_progress_m69679A44CDD90972865ADE60E388C1F7A68354B2 in Assembly-CSharp9.o
      _SessionStatus_get_ReadyForCreateProgress_mEC86C8F3156481CE60E50BD1C7991828DA9AEDDF in Assembly-CSharp9.o
     (maybe you meant: _NativeLibrary_ssc_session_status_get_ready_for_create_progress_m69679A44CDD90972865ADE60E388C1F7A68354B2)
  "_ssc_cloud_spatial_anchor_session_get_active_watchers_items", referenced from:
      _NativeLibrary_ssc_cloud_spatial_anchor_session_get_active_watchers_items_mB28130DA699E885EF779D87944E69436D42B6ED4 in Assembly-CSharp9.o
     (maybe you meant: _NativeLibrary_ssc_cloud_spatial_anchor_session_get_active_watchers_items_mB28130DA699E885EF779D87944E69436D42B6ED4)
  "_ssc_anchor_locate_criteria_set_near_device", referenced from:
      _NativeLibrary_ssc_anchor_locate_criteria_set_near_device_m384A852BF1FEFAC333F51635A355DEA3CF56AC25 in Assembly-CSharp9.o
     (maybe you meant: _NativeLibrary_ssc_anchor_locate_criteria_set_near_device_m384A852BF1FEFAC333F51635A355DEA3CF56AC25)
  "_ssc_cloud_spatial_anchor_session_set_log_level", referenced from:`

**Comment by clifflinmao** [comment]
> Sorry everyone. We're still working on getting an SDK out that makes everyone happy. We have several moving targets at play.
> 
> I have a branch ([craig/unity2019.3](https://github.com/Azure/azure-spatial-anchors-samples/tree/craig/unity2019.3)) that is all setup for 2019.3 using our existing SDK. It should work for all the platforms we currently support. I have not tried it on 2019.4 as that was only recently released. If anyone has a few minutes to try it out, please let me know.
> 
> Thanks!

I tried this branch with ARFoundation3, it works fine, but if I upgrade to ARFoundation 4, the camera is blank.

**Comment by craigktreasure** [comment]
We recently released [2.4.0](https://github.com/Azure/azure-spatial-anchors-samples/releases/tag/v2.4.0) with support for Unity 2019.4 (LTS). Please give the latest a shot and open new issues if needed.
